src/pudl/analysis/mcoe.py

```diff
@@ -38,7 +39,136 @@
+    ) -> pd.DataFrame:
+        return heat_rate_by_gen(bga=bga, hr_by_unit=hr_by_unit, gens=gens)
+
+    @asset(
+        name=f"fuel_cost_by_generator_{agg_freqs[freq]}",
+        ins={
+            "hr_by_gen": AssetIn(key=f"heat_rate_by_generator_{agg_freqs[freq]}"),
+            "gens": AssetIn(key="denorm_generators_eia"),
+            "frc": AssetIn(key=f"denorm_fuel_receipts_costs_{agg_freqs[freq]}_eia923"),
+        },
+        io_manager_key=io_manager_key,
+        compute_kind="Python",
+    )
+    def fc_asset(
+        hr_by_gen: pd.DataFrame, gens: pd.DataFrame, frc: pd.DataFrame
+    ) -> pd.DataFrame:
+        return fuel_cost(hr_by_gen=hr_by_gen, gens=gens, frc=frc)
+
+    @asset(
+        name=f"capacity_factor_by_generator_{agg_freqs[freq]}",
+        ins={
+            "gen": AssetIn(
+                key=f"generation_fuel_by_generator_{agg_freqs[freq]}_eia923"
+            ),
+            "gens": AssetIn(key="denorm_generators_eia"),
+        },
+        io_manager_key=io_manager_key,
+        compute_kind="Python",
+    )
+    def cf_asset(gens: pd.DataFrame, gen: pd.DataFrame) -> pd.DataFrame:
+        return capacity_factor(gens=gens, gen=gen, freq=freq)
+
+    @asset(
+        name=f"mcoe_{agg_freqs[freq]}",
+        ins={
+            "fuel_cost": AssetIn(key=f"fuel_cost_by_generator_{agg_freqs[freq]}"),
+            "capacity_factor": AssetIn(
+                key=f"capacity_factor_by_generator_{agg_freqs[freq]}"
+            ),
+        },
+        io_manager_key=io_manager_key,
+        compute_kind="Python",
+    )
+    def mcoe_asset(
+        fuel_cost: pd.DataFrame, capacity_factor: pd.DataFrame
+    ) -> pd.DataFrame:
+        return mcoe(
+            fuel_cost=fuel_cost,
+            capacity_factor=capacity_factor,
+            min_heat_rate=None,
+            min_fuel_cost_per_mwh=None,
+            min_cap_fact=None,
+            max_cap_fact=None,
+        )
+
+    @asset(
+        name=f"mcoe_generators_{agg_freqs[freq]}",
+        ins={
+            "mcoe": AssetIn(key=f"mcoe_{agg_freqs[freq]}"),
+            "gens": AssetIn(key="denorm_generators_eia"),
+        },
+        io_manager_key=io_manager_key,
+        compute_kind="Python",
+    )
+    def mcoe_generators_asset(mcoe: pd.DataFrame, gens: pd.DataFrame) -> pd.DataFrame:
+        return mcoe_generators(
+            mcoe=mcoe,
+            gens=gens,
+            freq=freq,
+            all_gens=True,
+            gens_cols=None,
+            timeseries_fillin=False,
+        )
+
+    return [
+        hr_by_unit_asset,
+        hr_by_gen_asset,
+        fc_asset,
+        cf_asset,
+        mcoe_asset,
+        mcoe_generators_asset,
+    ]
+
+
+mcoe_assets = [
+    mcoe_asset
+    for freq in ["AS", "MS"]
+    for mcoe_asset in mcoe_asset_factory(
+        freq=freq,
+        io_manager_key="pudl_sqlite_io_manager",
+    )
+]
+
+
+def heat_rate_by_unit(gen: pd.DataFrame, bf: pd.DataFrame) -> pd.DataFrame:
```

Comment:
Some of the docstring for this function seems obsolete, e.g. the parts talking about the BGA dataframe, which no longer appears to be involved?